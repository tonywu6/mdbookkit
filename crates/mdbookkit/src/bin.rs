#[cfg(feature = "lib-rustdoc-link")]
pub mod rustdoc_link;
