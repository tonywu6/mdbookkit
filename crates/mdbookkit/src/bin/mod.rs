#[cfg(feature = "lib-rustdoc-link")]
pub mod rustdoc_link;

#[cfg(feature = "lib-link-forever")]
pub mod link_forever;
