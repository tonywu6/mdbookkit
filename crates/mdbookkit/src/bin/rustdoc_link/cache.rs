use std::{
    borrow::Cow,
    collections::{BTreeSet, HashMap},
    hash::Hash,
    iter,
};

use anyhow::{bail, Context, Result};
use lsp_types::Url;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use sha2::{Digest, Sha256};
use tap::{Pipe, Tap, TapFallible};
use tokio::task::JoinSet;

use crate::log_debug;

use super::{env::Environment, link::ItemLinks, page::Pages, url::UrlToPath, Resolver};

#[allow(async_fn_in_trait)]
pub trait Cache: DeserializeOwned + Serialize {
    type Validated: Resolver;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated>;

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash;

    async fn load(env: &Environment) -> Result<Self::Validated> {
        env.load_temp::<Self, _>("cache.json")
            .tap_err(log_debug!())?
            .reuse(env)
            .await
            .tap_err(log_debug!())
    }

    async fn save<K>(env: &Environment, content: &Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash,
    {
        let this = Self::build(env, content).await?;
        env.save_temp::<Self, _>("cache.json", &this)
            .tap_err(log_debug!())
    }
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum FileCache {
    V1(FileCacheV1),
}

impl Cache for FileCache {
    type Validated = HashMap<String, ItemLinks>;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated> {
        match self {
            Self::V1(cache) => Ok(cache.reuse(env).await?),
        }
    }

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash,
    {
        Ok(Self::V1(FileCacheV1::build(env, content).await?))
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct FileCacheV1 {
    hash: String,
    urls: Vec<(String, ItemLinks)>,
}

impl Cache for FileCacheV1 {
    type Validated = HashMap<String, ItemLinks>;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated> {
        let deps = self
            .urls
            .iter()
            .flat_map(|(_, links)| links.deps())
            .map(Cow::Borrowed);

        let hash = Self::hash(env, deps).await;

        if hash != self.hash {
            bail!("checksum mismatch, expected {}, actual {hash}", self.hash)
        }

        Ok(self.urls.into_iter().collect())
    }

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash,
    {
        let urls = content
            .links()
            .iter()
            .map(|(k, v)| (k.to_string(), v.clone()))
            .collect::<Vec<_>>();

        let deps = urls
            .iter()
            .flat_map(|(_, links)| links.deps())
            .map(Cow::Borrowed);

        let hash = Self::hash(env, deps).await;

        Ok(Self { hash, urls })
    }
}

impl FileCacheV1 {
    async fn hash<'a, D>(env: &'a Environment, deps: D) -> String
    where
        D: Iterator<Item = Cow<'a, Url>>,
    {
        iter::once(Cow::Owned(env.source_dir.join("Cargo.toml").unwrap()))
            .chain(iter::once(Cow::Owned(
                env.crate_dir.join("Cargo.toml").unwrap(),
            )))
            .chain(iter::once(Cow::Borrowed(&env.entrypoint)))
            .chain(deps)
            .collect::<BTreeSet<_>>()
            .into_iter()
            .filter_map(|dep| {
                if env.source_dir.make_relative(&dep)?.starts_with("../") {
                    None
                } else {
                    Some(dep.into_owned())
                }
            })
            .map(read_dep)
            .collect::<JoinSet<_>>()
            .join_all()
            .await
            .into_iter()
            .filter_map(|result| {
                result
                    .context("failed to read cache dependency")
                    .tap_err(log_debug!())
                    .ok()
            })
            .collect::<Vec<_>>()
            .tap_mut(|deps| deps.sort_by(|(k1, _), (k2, _)| k1.cmp(k2)))
            .into_iter()
            .fold(Sha256::new(), |mut hash, (_, src)| {
                hash.update(src);
                hash
            })
            .pipe(|hash| hash.finalize().digest())
    }
}

async fn read_dep(url: Url) -> Result<(String, String)> {
    let content = tokio::fs::read_to_string(&url.to_path()?).await?;
    Ok((url.to_string(), content))
}

trait HexDigest {
    fn digest(&self) -> String;
}

impl HexDigest for sha2::digest::Output<Sha256> {
    fn digest(&self) -> String {
        format!("{:064x}", self)
    }
}
