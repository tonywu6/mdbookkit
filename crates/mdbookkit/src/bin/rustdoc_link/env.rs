use std::{
    ops::Deref,
    path::{Path, PathBuf},
    sync::Arc,
};

use anyhow::{anyhow, Context, Result};
use cargo_toml::{Manifest, Product};
#[cfg(feature = "common-cli")]
use clap::ValueHint;
use lsp_types::Url;
use pulldown_cmark::Options;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use shlex::Shlex;
use tap::Pipe;
use tokio::process::Command;

use crate::{env::ErrorHandling, markdown::mdbook_markdown};

use super::markdown;

#[derive(Deserialize, Debug, Default, Clone)]
#[cfg_attr(feature = "common-cli", derive(clap::Parser))]
#[serde(rename_all = "kebab-case", deny_unknown_fields)]
pub struct Config {
    /// Command to use for spawning rust-analyzer.
    ///
    /// By default, prebuilt binary from the VS Code extension is tried. If that doesn't
    /// exist, it is assumed that rust-analyzer is on `PATH`. Use this option to override
    /// this behavior completely.
    ///
    /// The command string will be tokenized by [shlex], so you can include arguments in it.
    #[serde(default)]
    #[cfg_attr(
        feature = "common-cli",
        arg(long, value_name("COMMAND"), value_hint(ValueHint::CommandString))
    )]
    pub rust_analyzer: Option<String>,

    /// List of features to activate when running rust-analyzer.
    ///
    /// This is just the `rust-analyzer.cargo.features` config.
    ///
    /// **In `book.toml`** — to enable all features, use `["all"]`.
    ///
    /// **For CLI** — to enable multiple features, specify as
    /// comma-separated values, or specify multiple times; to enable all features,
    /// specify `--cargo-features all`.
    #[cfg_attr(feature = "common-cli", arg(long, value_name("FEATURES")))]
    #[serde(default)]
    pub cargo_features: Vec<String>,

    /// Directory from which to search for a Cargo project.
    ///
    /// By default, the current working directory is used. Use this option to specify a
    /// different directory.
    ///
    /// The processor requires the Cargo.toml of a package to work. If you are working
    /// on a Cargo workspace, set this to the relative path to a member crate.
    #[cfg_attr(
        feature = "common-cli",
        arg(long, value_name("PATH"), value_hint(ValueHint::DirPath))
    )]
    #[serde(default)]
    pub manifest_dir: Option<PathBuf>,

    /// Directory in which to persist build cache.
    ///
    /// Setting this will enable caching. Will skip rust-analyzer if cache hits.
    #[cfg_attr(
        feature = "common-cli",
        arg(long, value_name("PATH"), value_hint(ValueHint::DirPath))
    )]
    #[serde(default)]
    pub cache_dir: Option<PathBuf>,

    /// Whether to exit with failure when some links fail to resolve.
    ///
    /// Warnings are always emitted for unresolved links regardless of this option.
    #[serde(default)]
    #[cfg_attr(feature = "common-cli", arg(long, value_enum, value_name("MODE"), default_value_t = Default::default()))]
    pub fail_on_unresolved: ErrorHandling,

    /// Whether to enable punctuations like smart quotes `“”`.
    ///
    /// This is only meaningful if your links happen to have visible text that has
    /// specific punctuation. The processor otherwise passes through the rest of your
    /// Markdown source.
    ///
    /// **In `book.toml`** — this option is not needed because
    /// `output.html.smart-punctuation` is honored.
    #[serde(default)]
    #[cfg_attr(feature = "common-cli", arg(long))]
    pub smart_punctuation: bool,

    #[serde(default)]
    #[cfg_attr(feature = "common-cli", arg(long, hide = true))]
    pub prefer_local_links: bool,

    #[allow(unused)]
    #[serde(default)]
    #[doc(hidden)]
    #[cfg_attr(feature = "common-cli", arg(skip))]
    pub after: Option<Vec<String>>,

    #[allow(unused)]
    #[serde(default)]
    #[doc(hidden)]
    #[cfg_attr(feature = "common-cli", arg(skip))]
    pub before: Option<Vec<String>>,

    #[allow(unused)]
    #[serde(default)]
    #[doc(hidden)]
    #[cfg_attr(feature = "common-cli", arg(skip))]
    pub renderers: Option<Vec<String>>,

    #[allow(unused)]
    #[serde(default)]
    #[doc(hidden)]
    #[cfg_attr(feature = "common-cli", arg(skip))]
    pub command: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Environment {
    pub temp_dir: TempDir,
    pub crate_dir: Url,
    pub source_dir: Url,
    pub entrypoint: Url,
    pub config: Config,
}

impl Environment {
    pub fn new(config: Config) -> Result<Self> {
        let cwd = config
            .manifest_dir
            .clone()
            .map(Ok)
            .unwrap_or_else(std::env::current_dir)?
            .canonicalize()?;

        let (crate_dir, entrypoint) = {
            let manifest_path = LocateProject::package(&cwd)?.root;

            let manifest = {
                let mut manifest = Manifest::from_path(&manifest_path)?;
                manifest.complete_from_path(&manifest_path)?;
                manifest
            };

            let crate_dir = manifest_path
                .parent()
                .unwrap()
                .pipe(Url::from_directory_path)
                .unwrap();

            if let Some(Product {
                path: Some(ref lib),
                ..
            }) = manifest.lib
            {
                let entry = crate_dir.join(lib)?;
                Ok((crate_dir, entry))
            } else if let Some(bin) = manifest.bin.iter().find_map(|bin| bin.path.as_ref()) {
                let entry = crate_dir.join(bin)?;
                Ok((crate_dir, entry))
            } else {
                Err(anyhow!(
                    "help: resolved Cargo.toml is {}",
                    manifest_path.display()
                ))
                .pipe(|r| {
                    if manifest.workspace.is_some() {
                        r.context("help: to use in a workspace, set `manifest-dir` option to root of a member crate")
                    } else {
                        r
                    }
                })
                .context("Cargo.toml does not have any lib or bin target")
            }
        }?;

        let source_dir = LocateProject::workspace(cwd)?
            .root
            .parent()
            .unwrap()
            .pipe(Url::from_directory_path)
            .unwrap();

        let temp_dir = config
            .cache_dir
            .clone()
            .map(TempDir::Persistent)
            .or_else(|| {
                tempfile::TempDir::new()
                    .ok()
                    .map(Arc::new)
                    .map(TempDir::Transient)
            })
            .context("failed to obtain a temporary directory")?;

        Ok(Self {
            temp_dir,
            crate_dir,
            source_dir,
            entrypoint,
            config,
        })
    }

    pub fn command(&self) -> Result<Command> {
        if let Some(command) = self.config.rust_analyzer.as_deref() {
            let mut words = Shlex::new(command);
            let executable = words
                .next()
                .context("unexpected empty string for option `rust-analyzer`")?;
            let mut cmd = Command::new(executable);
            cmd.args(words);
            Ok(cmd)
        } else if let Some(extension) = find_code_extension() {
            log::debug!("using rust-analyzer from {}", extension.display());
            Ok(Command::new(extension))
        } else {
            Ok(Command::new("rust-analyzer"))
        }
    }

    pub fn markdown<'a>(&self, source: &'a str) -> markdown::MarkdownStream<'a> {
        let options = if self.config.smart_punctuation {
            Options::ENABLE_SMART_PUNCTUATION
        } else {
            Options::empty()
        };
        markdown::stream(source, options.union(mdbook_markdown()))
    }

    pub fn emit_config(&self) -> EmitConfig {
        let Config {
            prefer_local_links, ..
        } = self.config;
        EmitConfig { prefer_local_links }
    }

    pub fn load_temp<T, P>(&self, path: P) -> Result<T>
    where
        T: DeserializeOwned,
        P: AsRef<Path>,
    {
        let path = self.temp_dir.as_ref().join(path);
        let text = std::fs::read_to_string(&path).context("failed to read from cache dir")?;
        Ok(serde_json::from_str(&text)?)
    }

    pub fn save_temp<T, P>(&self, path: P, temp: &T) -> Result<()>
    where
        T: Serialize,
        P: AsRef<Path>,
    {
        let path = self.temp_dir.as_ref().join(path);
        let text = serde_json::to_string(&temp).context("failed to serialize cache data")?;
        std::fs::create_dir_all(path.parent().unwrap()).context("failed to create cache dir")?;
        std::fs::write(path, text).context("failed to write to cache dir")?;
        Ok(())
    }
}

pub struct EmitConfig {
    pub prefer_local_links: bool,
}

#[derive(Debug, Clone)]
pub enum TempDir {
    Persistent(PathBuf),
    Transient(Arc<tempfile::TempDir>),
}

impl AsRef<Path> for TempDir {
    fn as_ref(&self) -> &Path {
        match self {
            Self::Persistent(p) => p.as_ref(),
            Self::Transient(p) => p.deref().as_ref(),
        }
    }
}

#[derive(Deserialize)]
struct LocateProject {
    root: PathBuf,
}

impl LocateProject {
    fn package<P: AsRef<Path>>(cwd: P) -> Result<Self> {
        std::process::Command::new(env!("CARGO"))
            .arg("locate-project")
            .arg("--message-format=json")
            .current_dir(cwd)
            .output()?
            .pipe(Self::parse)
    }

    fn workspace<P: AsRef<Path>>(cwd: P) -> Result<Self> {
        std::process::Command::new(env!("CARGO"))
            .arg("locate-project")
            .arg("--message-format=json")
            .arg("--workspace")
            .current_dir(cwd)
            .output()?
            .pipe(Self::parse)
    }

    fn parse(output: std::process::Output) -> Result<Self> {
        if output.status.success() {
            String::from_utf8(output.stdout)?
                .pipe(|outout| serde_json::from_str::<Self>(&outout))?
                .pipe(Ok)
        } else {
            anyhow!(String::from_utf8_lossy(&output.stderr).into_owned())
                .context("help: a Cargo project is needed to run rust-analyzer in")
                .context("failed to locate a Cargo project")
                .pipe(Err)
        }
    }
}

fn find_code_extension() -> Option<PathBuf> {
    let home = dirs::home_dir()?;
    [
        home.join(".vscode/extensions"),
        home.join(".vscode-server/extensions"),
    ]
    .iter()
    .flat_map(|p| p.read_dir().ok())
    .flatten()
    .flatten()
    .find_map(|extension| {
        if extension
            .file_name()
            .to_string_lossy()
            .starts_with("rust-lang.rust-analyzer-")
        {
            Some(extension.path().join("server/rust-analyzer"))
        } else {
            None
        }
    })
}
