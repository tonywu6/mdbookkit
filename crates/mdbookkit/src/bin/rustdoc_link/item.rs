use syn::{
    parenthesized,
    parse::{End, Parse, ParseStream, Parser},
    spanned::Spanned,
    token::Paren,
    PathArguments, QSelf, Token, TypePath,
};

#[derive(Debug)]
pub struct Item {
    pub name: String,
    pub stmt: String,
    pub cursor: Cursor,
}

impl Item {
    pub fn parse(path: &str) -> anyhow::Result<Self> {
        let path = match path.split_once('@') {
            None => path,
            Some((_, path)) => path,
        };

        let item = ItemName::parse.parse_str(path)?;

        let (name, column) = {
            let mut name = String::new();
            let mut column = 0;

            let gt = if let Some(QSelf { ty, position, .. }) = item.path.qself {
                name.push('<');
                name.push_str(&path[ty.span().byte_range()]);
                name.push_str(" as ");
                Some(position - 1)
            } else {
                None
            };

            for (idx, chunk) in item.path.path.segments.pairs().enumerate() {
                column = name.len();

                let leading = &chunk.value().ident.span();

                name.push_str(&path[leading.span().byte_range()]);

                match &chunk.value().arguments {
                    PathArguments::None => {}

                    PathArguments::AngleBracketed(args) => {
                        if args.colon2_token.is_none() {
                            // make it a turbofish
                            name.push_str("::");
                        }
                        name.push_str(&path[args.span().byte_range()])
                    }

                    PathArguments::Parenthesized(args) => {
                        name.push_str(&path[args.span().byte_range()])
                    }
                }

                if gt == Some(idx) {
                    name.push('>');
                }

                if let Some(punct) = chunk.punct() {
                    name.push_str(&path[punct.span().byte_range()]);
                }
            }

            (name, column)
        };

        let (stmt, cursor) = match item.kind {
            None => {
                let pattern = "let _: ";
                let assign = " = ";
                let stmt = format!("{pattern}{name}{assign}{name};");

                let c1 = pattern.len() + column;
                let c2 = pattern.len() + name.len() + assign.len() + column;
                let cursor = Cursor::Decl([c1, c2]);

                (stmt, cursor)
            }
            Some(ItemKind::Call) => (format!("{name}();"), Cursor::Expr([column])),
            Some(ItemKind::Macro) => (format!("{name}!();"), Cursor::Expr([column])),
        };

        Ok(Self { name, stmt, cursor })
    }
}

#[derive(Debug)]
pub enum Cursor {
    Decl([usize; 2]),
    Expr([usize; 1]),
}

impl AsRef<[usize]> for Cursor {
    fn as_ref(&self) -> &[usize] {
        match self {
            Self::Decl(c) => c,
            Self::Expr(c) => c,
        }
    }
}

struct ItemName {
    path: TypePath,
    kind: Option<ItemKind>,
}

#[derive(Debug, Clone, Copy)]
enum ItemKind {
    Call,
    Macro,
}

impl Parse for ItemName {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let path = input.parse::<TypePath>()?;
        if input.peek(Token![!]) {
            // as a macro
            input.parse::<Token![!]>()?;
            input.step(|cursor| {
                if let Some((inner, _, _, rest)) = cursor.any_group() {
                    if inner.eof() {
                        Ok(((), rest))
                    } else {
                        Err(cursor.error("expected no arguments"))
                    }
                } else {
                    Ok(((), *cursor))
                }
            })?;
            input.parse::<Eof>()?;
            Ok(Self {
                path,
                kind: Some(ItemKind::Macro),
            })
        } else if input.peek(Paren) {
            // as a function
            let inner;
            let _ = parenthesized!(inner in input);
            if inner.is_empty() {
                Ok(Self {
                    path,
                    kind: Some(ItemKind::Call),
                })
            } else {
                Err(input.error("expected no arguments"))
            }
        } else {
            // as a path
            input.parse::<Eof>()?;
            Ok(Self { path, kind: None })
        }
    }
}

struct Eof;

impl Parse for Eof {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        if input.peek(End) {
            Ok(Self)
        } else {
            Err(input.error("expected no additional token"))
        }
    }
}
