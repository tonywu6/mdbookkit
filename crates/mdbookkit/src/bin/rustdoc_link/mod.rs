use std::{borrow::Borrow, collections::HashMap, hash::Hash, sync::Arc};

use anyhow::{Context, Result};
use lsp_types::Position;
use tap::{Pipe, TapFallible};
use tokio::task::JoinSet;

mod client;
pub mod env;
mod item;
mod link;
mod markdown;
mod page;
mod sync;

#[cfg(feature = "rustdoc-link")]
pub mod cache;

use crate::{log_debug, logging::spinner, styled};

pub use self::{client::Client, page::Pages};
use self::{item::Item, link::ItemLinks};

/// Type that can provide links.
///
/// Resolvers should modify the provided [`Pages`] in place.
///
/// This is currently an abstraction over two sources of links:
///
/// - [`Client`], which invokes rust-analyzer
/// - [`Cache`][crate::cache::Cache] implementations
#[allow(async_fn_in_trait)]
pub trait Resolver {
    async fn resolve<K>(&self, pages: &mut Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash;
}

impl Resolver for Client {
    async fn resolve<K>(&self, pages: &mut Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash,
    {
        let request = pages.items();

        if request.is_empty() {
            return Ok(());
        }

        let main = std::fs::read_to_string(self.env().entrypoint.path())?;

        let (context, request) = {
            let mut context = format!("{main}\nfn {UNIQUE_ID} () {{\n");

            let line = context.chars().filter(|&c| c == '\n').count();

            let request = request
                .iter()
                .scan(line, |line, (key, item)| {
                    build(&mut context, line, item).map(|cursors| (key.clone(), cursors))
                })
                .collect::<Vec<_>>();

            fn build(context: &mut String, line: &mut usize, item: &Item) -> Option<Vec<Position>> {
                use std::fmt::Write;
                let _ = writeln!(context, "{}", item.stmt);
                let cursors = item
                    .cursor
                    .as_ref()
                    .iter()
                    .map(|&col| Position::new(*line as _, col as _))
                    .collect::<Vec<_>>();
                *line += 1;
                Some(cursors)
            }

            context.push('}');

            (context, request)
        };

        log::debug!("request context\n\n{context}\n");

        let document = self
            .open(self.env().entrypoint.clone(), context)
            .await?
            .pipe(Arc::new);

        spinner().create("resolve", Some(request.len() as _));

        let tasks: JoinSet<Option<(String, ItemLinks)>> = request
            .into_iter()
            .map(|(key, pos)| {
                let key = key.to_string();
                let doc = document.clone();
                resolve(doc, key, pos)
            })
            .collect();

        async fn resolve(
            doc: Arc<client::OpenDocument>,
            key: String,
            pos: Vec<Position>,
        ) -> Option<(String, ItemLinks)> {
            let _task = spinner().task("resolve", &key);
            for p in pos {
                let resolved = doc
                    .resolve(p)
                    .await
                    .with_context(|| format!("error resolving {p:?}"))
                    .tap_err(log_debug!())
                    .ok();
                if let Some(resolved) = resolved {
                    return Some((key, resolved));
                }
            }
            None
        }

        let resolved = tasks
            .join_all()
            .await
            .into_iter()
            .flatten()
            .collect::<HashMap<_, _>>();

        spinner().finish("resolve", styled!(("done").green()));

        pages.apply(&resolved);

        Ok(())
    }
}

impl<K> Resolver for HashMap<K, ItemLinks>
where
    K: Borrow<str> + Eq + Hash,
{
    async fn resolve<P>(&self, pages: &mut Pages<'_, P>) -> Result<()>
    where
        P: Eq + Hash,
    {
        pages.apply(self);
        Ok(())
    }
}

const UNIQUE_ID: &str = "__ded48f4d_0c4f_4950_b17d_55fd3b2a0c86__";
