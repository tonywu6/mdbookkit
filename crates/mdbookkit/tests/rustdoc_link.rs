use std::{io::Write, sync::Arc};

use anyhow::{bail, Context, Result};

use assert_cmd::{prelude::*, Command};
use predicates::prelude::*;
use similar::{ChangeTag, TextDiff};
use tap::Pipe;
use tempfile::TempDir;
use tokio::task::JoinSet;

use mdbookkit::bin::rustdoc_link::{
    env::{find_code_extension, Config, Environment},
    Client, Pages, Resolver,
};
use util_testing::{may_skip, portable_snapshots, setup_paths, test_document, TestDocument};

mod util;

async fn snapshot(
    client: Arc<Client>,
    TestDocument { source, name, .. }: TestDocument,
) -> Result<()> {
    let stream = client.env().markdown(source).into_offset_iter();

    let mut page = Pages::one(source, stream)?;

    client.resolve(&mut page).await?;

    let output = page.get(&client.env().emit_config())?.to_string();

    portable_snapshots!().test(|| insta::assert_snapshot!(name.clone(), output))?;

    assert_no_whitespace_change(source, &output)?;

    let report = page
        .reporter()
        .level(log::LevelFilter::Info)
        .names(|_| name.clone())
        .colored(false)
        .logging(false)
        .build()
        .to_report();

    portable_snapshots!().test(|| insta::assert_snapshot!(format!("{name}.stderr"), report))?;

    Ok(())
}

fn assert_no_whitespace_change(source: &str, output: &str) -> Result<()> {
    let changed_lines = TextDiff::from_words(source, output)
        .iter_all_changes()
        .filter_map(|change| {
            if matches!(change.tag(), ChangeTag::Equal) {
                return None;
            }
            if change.value().contains('\n') {
                Some(change.value())
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    if !changed_lines.is_empty() {
        bail!("unexpected whitespace change: {changed_lines:?}")
    } else {
        Ok(())
    }
}

#[tokio::test]
async fn test_snapshots() -> Result<()> {
    util::setup_logging();

    let client = client()?;

    let tests = [
        test_document!("../../../docs/src/rustdoc-link/supported-syntax.md"),
        test_document!("../../../docs/src/rustdoc-link/known-issues.md"),
        test_document!("../../../docs/src/rustdoc-link/getting-started.md"),
        test_document!("../../../docs/src/rustdoc-link/index.md"),
        test_document!("tests/ra-known-quirks.md"),
    ];

    let errors = tests
        .map(|test| snapshot(client.clone(), test))
        .into_iter()
        .collect::<JoinSet<_>>()
        .join_all()
        .await
        .into_iter()
        .filter_map(Result::err)
        .collect::<Vec<_>>();

    if !errors.is_empty() {
        let errors = errors
            .iter()
            .map(|e| format!("{e:?}"))
            .collect::<Vec<_>>()
            .join("\n");
        panic!("{errors}")
    }

    client.drop().await?;

    Ok(())
}

fn client() -> Result<Arc<Client>> {
    Config {
        rust_analyzer: Some("cargo run --package util-rust-analyzer -- analyzer".into()),
        cargo_features: vec!["rustdoc-link".into()],
        ..Default::default()
    }
    .pipe(Environment::new)?
    .pipe(Client::new)
    .pipe(Arc::new)
    .pipe(Ok)
}

#[test]
#[ignore = "should run in a dedicated environment"]
fn test_minimum_env() -> Result<()> {
    util::setup_logging();

    log::info!("setup: compile self");
    Command::new("cargo")
        .args([
            "build",
            "--package",
            env!("CARGO_PKG_NAME"),
            "--all-features",
            "--bin",
            "mdbook-rustdoc-link",
        ])
        .arg(if cfg!(debug_assertions) {
            "--profile=dev"
        } else {
            "--profile=release"
        })
        .assert()
        .success();

    let path = setup_paths()?;

    let root = TempDir::new()?;

    log::debug!("{root:?}");

    log::info!("given: a book");
    Command::new("mdbook")
        .args(["init", "--force"])
        .env("PATH", &path)
        .current_dir(&root)
        .unwrap()
        .assert()
        .success();

    log::info!("given: preprocessor is enabled");
    std::fs::File::options()
        .append(true)
        .open(root.path().join("book.toml"))?
        .pipe(|mut file| file.write_all("[preprocessor.rustdoc-link]\n".as_bytes()))?;

    log::info!("when: book is not a Cargo project");
    log::info!("then: preprocessor fails");
    Command::new("mdbook")
        .arg("build")
        .env("PATH", &path)
        .current_dir(&root)
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "failed to determine the current Cargo project",
        ));

    log::info!("given: book is a Cargo project");
    Command::new("cargo")
        .arg("init")
        .args(["--name", "temp"])
        .env("PATH", &path)
        .current_dir(&root)
        .assert()
        .success();

    if find_code_extension().is_some()
        && may_skip("rust-analyzer code extension is already installed")
    {
        log::info!("when: book has item links");
        std::fs::File::options()
            .append(true)
            .open(root.path().join("src/chapter_1.md"))?
            .pipe(|mut file| file.write_all("\n[std::thread]\n".as_bytes()))?;

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();
    } else if Command::new("rust-analyzer")
        .arg("--version")
        .assert()
        .try_success()
        .is_ok()
        && may_skip("rust-analyzer is already available")
    {
        log::info!("skip testing mdbook build without rust-analyzer")
    } else {
        log::info!("when: rust-analyzer is not configured");

        log::info!("when: book has no item links");

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();

        log::info!("when: book has item links");
        std::fs::File::options()
            .append(true)
            .open(root.path().join("src/chapter_1.md"))?
            .pipe(|mut file| file.write_all("\n[std]\n".as_bytes()))?;

        log::info!("then: preprocessor fails");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .failure()
            .stderr(
                predicate::str::contains("failed to spawn rust-analyzer")
                    // https://github.com/rust-lang/rustup/issues/3846
                    // rustup shims rust-analyzer when it's not installed
                    .or(predicate::str::contains("Unknown binary 'rust-analyzer")),
                // ^ doesn't have a closing `'` because on windows it says 'rust-analyzer.exe'
            );

        log::info!("when: code extension is installed");

        let extension_dir = tempfile::Builder::new()
            .prefix(".vscode")
            .suffix("")
            .rand_bytes(0)
            .tempdir_in(dirs::home_dir().context("failed to get home dir")?)?;

        let ra_executable = extension_dir
            .path()
            .join("extensions/rust-lang.rust-analyzer-lorem-ipsum")
            .join("server/rust-analyzer");

        Command::new("cargo")
            .args(["run", "--package", "util-rust-analyzer", "--"])
            .arg("--ra-path")
            .arg(ra_executable)
            .arg("download")
            .unwrap()
            .assert()
            .success();

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();
    }

    Ok(())
}
