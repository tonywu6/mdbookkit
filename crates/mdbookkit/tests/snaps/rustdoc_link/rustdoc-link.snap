---
source: crates/mdbookkit/tests/rustdoc_link.rs
expression: output
---
# mdbook-rustdoc-link

**[_rustdoc_-style linking][rustdoc] for [mdBook]** (with the help of [rust-analyzer]).

You write:

```md
The [`option`][std::option] and [`result`][std::result] modules define optional and
error-handling types, [`Option<T>`] and [`Result<T, E>`]. The [`iter`][std::iter] module
defines Rust's iterator trait, [`Iterator`], which works with the `for` loop to access
collections. [^1]
```

You get:

<figure class="fig-text">

The [`option`](https://doc.rust-lang.org/stable/core/option/index.html "std::option") and [`result`](https://doc.rust-lang.org/stable/core/result/index.html "std::result") modules define optional and
error-handling types, [`Option<T>`](https://doc.rust-lang.org/stable/core/option/enum.Option.html "Option<T>") and [`Result<T, E>`](https://doc.rust-lang.org/stable/core/result/enum.Result.html "Result<T, E>"). The [`iter`](https://doc.rust-lang.org/stable/core/iter/index.html "std::iter") module
defines Rust's iterator trait, [`Iterator`](https://doc.rust-lang.org/stable/core/iter/traits/iterator/trait.Iterator.html "Iterator"), which works with the `for` loop to access
collections. [^1]

</figure>

`mdbook-rustdoc-link` is an mdBook [preprocessor]. Using rust-analyzer, it converts type
names, module paths, etc. into links to online crate docs. No more finding and pasting
URLs by hand.

<figure>

![screen recording of mdbook-rustdoc-link during mdbook build](rustdoc-link/media/screencap.webp)

</figure>

## Overview

To get started, simply follow the [quickstart guide](rustdoc-link/getting-started.md)!

If you would like to read more about this crate:

For **writing documentation** —

- To know more about how the preprocessor is resolving items into links, see
  [Name resolution](rustdoc-link/name-resolution.md).
- To know how to link to other types of items like
  [functions, macros](rustdoc-link/supported-syntax.md#functions-and-macros), and
  [implementors](rustdoc-link/supported-syntax.md#implementors-and-fully-qualified-syntax),
  see [Supported syntax](rustdoc-link/supported-syntax.md).

For **adapting this crate to your project** —

- If you use [Cargo workspaces][workspaces], see specific instructions in
  [Workspace layout](rustdoc-link/workspace-layout.md).
- If you are working on a large project, and processing is taking a long time, see the
  discussion in [Caching](rustdoc-link/caching.md).

For **additional usage information** —

- You can use this as a standalone command line tool: see
  [Standalone usage](rustdoc-link/standalone-usage.md).
- For all available options and how to set them, see
  [Configuration](rustdoc-link/configuration.md).
- Finally, review [Known issues](rustdoc-link/known-issues.md) and limitations.

Happy linking!

---

[^1]: Text adapted from [<cite>A Tour of The Rust Standard Library</cite>][tour]

<!-- prettier-ignore-start -->

[mdBook]: https://rust-lang.github.io/mdBook/
[preprocessor]: https://rust-lang.github.io/mdBook/format/configuration/preprocessors.html
[rust-analyzer]: https://rust-analyzer.github.io/
[rustdoc]: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html
[tour]: https://doc.rust-lang.org/stable/std/#a-tour-of-the-rust-standard-library
[workspaces]: https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html

<!-- prettier-ignore-end -->
